polkadot/node/core/approval-voting/src/criteria.rs

```diff
@@ -88,25 +96,105 @@ impl From<OurAssignment> for crate::approval_db::v1::OurAssignment {
-	// combine the relay VRF story with a sample number.
-	let mut t = Transcript::new(approval_types::RELAY_VRF_MODULO_CONTEXT);
-	t.append_message(b"RC-VRF", &relay_vrf_story.0);
-	sample.using_encoded(|s| t.append_message(b"sample", s));
+// Combines the relay VRF story with a sample number if any.
+fn relay_vrf_modulo_transcript_inner(
+	mut transcript: Transcript,
+	relay_vrf_story: RelayVRFStory,
+	sample: Option<u32>,
+) -> Transcript {
+	transcript.append_message(b"RC-VRF", &relay_vrf_story.0);
 
-	t
+	if let Some(sample) = sample {
+		sample.using_encoded(|s| transcript.append_message(b"sample", s));
+	}
+
+	transcript
+}
+
+fn relay_vrf_modulo_transcript_v1(relay_vrf_story: RelayVRFStory, sample: u32) -> Transcript {
+	relay_vrf_modulo_transcript_inner(
+		Transcript::new(approval_types::v1::RELAY_VRF_MODULO_CONTEXT),
+		relay_vrf_story,
+		Some(sample),
+	)
+}
+
+fn relay_vrf_modulo_transcript_v2(relay_vrf_story: RelayVRFStory) -> Transcript {
+	relay_vrf_modulo_transcript_inner(
+		Transcript::new(approval_types::v2::RELAY_VRF_MODULO_CONTEXT),
+		relay_vrf_story,
+		None,
+	)
+}
+
+/// A hard upper bound on num_cores * target_checkers / num_validators
+const MAX_MODULO_SAMPLES: usize = 40;
+
+/// Takes the VRF output as input and returns a Vec of cores the validator is assigned
+/// to as a tranche0 checker.
+fn relay_vrf_modulo_cores(
+	vrf_in_out: &VRFInOut,
+	// Configuration - `relay_vrf_modulo_samples`.
+	num_samples: u32,
+	// Configuration - `n_cores`.
+	max_cores: u32,
+) -> Vec<CoreIndex> {
+	let rand_chacha =
+		ChaCha20Rng::from_seed(vrf_in_out.make_bytes::<<ChaCha20Rng as SeedableRng>::Seed>(
+			approval_types::v2::CORE_RANDOMNESS_CONTEXT,
+		));
+	generate_samples(rand_chacha, num_samples as usize, max_cores as usize)
+}
+
+/// Generates `num_sumples` randomly from (0..max_cores) range
+///
+/// Note! The algorithm can't change because validators on the other
+/// side won't be able to check the assignments until they update.
+/// This invariant is tested with `generate_samples_invariant`, so the
+/// tests will catch any subtle changes in the implementation of this function
+/// and its dependencies.
+fn generate_samples(
+	mut rand_chacha: ChaCha20Rng,
+	num_samples: usize,
+	max_cores: usize,
+) -> Vec<CoreIndex> {
+	if num_samples as usize > MAX_MODULO_SAMPLES {
+		gum::warn!(
+			target: LOG_TARGET,
+			n_cores = max_cores,
+			num_samples,
+			max_modulo_samples = MAX_MODULO_SAMPLES,
+			"`num_samples` is greater than `MAX_MODULO_SAMPLES`",
+		);
+	}
+
+	if 2 * num_samples > max_cores {
+		gum::debug!(
+			target: LOG_TARGET,
+			n_cores = max_cores,
+			num_samples,
+			max_modulo_samples = MAX_MODULO_SAMPLES,
+			"Suboptimal configuration `num_samples` should be less than `n_cores` / 2",
+		);
+	}
+
+	let num_samples = min(MAX_MODULO_SAMPLES, min(num_samples, max_cores));
+
+	let mut random_cores = (0..max_cores as u32).map(|val| val.into()).collect::<Vec<CoreIndex>>();
```

Comment:
this could be restricted to core_ids which have actually concluded availability

Reply:
This would make the number of tranche0 checks deterministic wrt number of available candidates. However this would have the effect of doing way more checks than needed if there are few candidates that become available.